File: Algorithms/C++/00128-Longest Consecutive Sequence.cpp
class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        std::unordered_set<int> set;

        // populate set 
        for(int i{0}; i < nums.size(); i++){
            set.insert(nums[i]);
        }

        int longestSeq{0};
        for(int i{0}; i < nums.size(); i++){
            if(!set.contains(nums[i]-1)){
                int next = nums[i] + 1;
                int currSeq{1};
                while(set.contains(next)){
                    currSeq++;
                    next++;
                }
                longestSeq = std::max(longestSeq, currSeq);
            }
        }

        return longestSeq;
    }
};

/*
    - cant sort
    - cant use sorted data structure
    - return length of longest consecutive sequence

    algorithm
    - build a set from nums
    - check if number starts a segment (no left neighbor)
    - build segment by finding next values from set
    - take max of longestSeq and currSeq
    - return longestSeq

    [1,2,3,4] --- [100] --- [200]

*/

File: Algorithms/Python3/00128-Longest Consecutive Sequence.py
class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        if not nums:
            return 0

        hashset = set(nums)
        longestConsecutive = 1

        for num in hashset:
            consecutive = 1
            if num-1 not in hashset:
                num += 1 
                while(num in hashset):
                    consecutive += 1
                    num += 1

            longestConsecutive = max(longestConsecutive, consecutive)

        return longestConsecutive

'''
hashset
(100, 4, 200, 1, 3, 2)

check if num is FIRST in subsequence
if not first, skip
if first, check length of subsequence

100
check for 101
4
check for 5
200
check for 201
1
check for 2
check for 3
check for 4 

return max consecutive elems

tc: O(n + n + max cons. elems)
sc: O(n)
'''